jax/random.py

```diff
@@ -946,6 +946,24 @@ def _gumbel(key, shape, dtype):
   return -np.log(-np.log(
       uniform(key, shape, dtype, minval=np.finfo(dtype).eps, maxval=1.)))
 
+def categorical(key, logits, axis=-1, shape=()):
+  """Sample random values from categorical distributions.
+
+  Args:
+    key: a PRNGKey used as the random key.
+    logits: Unnormalized log probabilities of the categorical distribution(s) to sample from.
+    axis: Specifies the axis along which logits belong to the same categorical distribution.
+    shape: Specifies how many samples to take per categorical distribution and in what shape to arrange them.
+
+  Returns:
+    Samples of shape `shape + tuple(onp.delete(logits.shape, axis))`.
+    The first len(shape) dimensions specify which sample,
+    while the last len(logits.shape) - 1 dimensions make up the shape per sample.
+    """
+  if axis >= 0:
```

Comment:
That should be `<`, not `>=` ;-) But also I don't think you need to touch `axis` at all since you just forward it to `argmax`.

Reply:
It has to be `>=` since the axis dimension has to be realigned to the logit dimensions which can be achieved by skipping `len(shape)` dimensions. This is not necessary for negative arguments.

Reply:
Whoops, you're right.